### ISSUE TITLE
Inference on the tinygo webassembly version not working correctly

### ISSUE THREAD

**Issue by gc-2412** [hybridgroup/yzma#341]
Inference doesn't appear to be working properly on the [webpage](https://yzma.ai/try/) . 

<img width="621" height="910" alt="Image" src="https://github.com/user-attachments/assets/b01e4e73-d661-4a39-9073-8371a8aa9e53" />

**Comment by deadprogram** [comment]
Please provide the browser/OS version you are using @gc-2412 thanks.

**Comment by gc-2412** [comment]
OS: Ubuntu 22.04.5 LTS
Browser: Google Chrome Version 152.0.7977.82 (Official Build) (x86_64)

**Comment by deadprogram** [comment]
Thank you for the version information.

I cannot reproduce this. I ran the same model (unsloth/Qwen3.5-0.8B-GGUF
Q4_K_M) on the llama.cpp build that the page used on 11 September, on the
single thread build and on the thread build at 2, 4, 8, and 16 threads.
Every run gave a correct answer.

Two questions.

1. The page prints the backend at the top right, above the Model list. Your
   picture does not show that line. What does it say?

2. Please open https://yzma.ai/try/app/?mode=cpu and ask the same question.
   This makes the page use the CPU. If the answer is correct there, the
   problem is in the WebGPU backend of llama.cpp on your machine.

Thanks!

**Comment by gc-2412** [comment]
You might be on to something. 
I tried it again today on  https://yzma.ai/try/  and got a similar result. 
[full screen shot]
<img width="1761" height="1035" alt="Image" src="https://github.com/user-attachments/assets/c948b7b7-5a3e-48f1-a253-cc993bcbb888" />

and this one on https://yzma.ai/try/app/?mode=cpu  gave a much better response

<img width="1809" height="1032" alt="Image" src="https://github.com/user-attachments/assets/a3d66caa-75cf-4821-8fd6-c8b611dde9c3" />

**Comment by deadprogram** [comment]
Thank you, that answers it. The two pictures show the cause.

- `yzma.ai/try/` says `backend: webgpu (WebGPU)` and gives nonsense.
- `?mode=cpu` says `backend: cpu-threads, 8 threads` and gives a correct
  answer.

The same WebAssembly build and the same model give both results, so the model
and the yzma code are not at fault. The WebGPU backend of llama.cpp computes
wrong values on your machine. The text holds pieces of many languages, which
is what greedy sampling gives when the logits are close to noise and each step
takes a random token of the vocabulary.

We know this failure from Firefox, where WebGPU stops the answer at the first
token. Auto mode already takes the CPU there. Your report is the first one on
Chrome on Linux, and the text goes on instead of stopping, so this is a
different driver path.

Please use `https://yzma.ai/try/app/?mode=cpu` until this has a repair.

One more request, so that we can find the driver that does this. Open
`chrome://gpu` and give us these three lines.

1. The `GL_RENDERER` line, which names the card.
2. The `Vulkan` line of the Graphics Feature Status list.
3. The first adapter line of Dawn Info, which says `Vulkan backend` or
   `OpenGLES backend`.

A picture of the top of that page is sufficient.

**Comment by gc-2412** [comment]
Top of the page

<img width="500" height="328" alt="Image" src="https://github.com/user-attachments/assets/be01d65d-e58d-408b-81c9-431284f0cde4" />



`GL_RENDERER  : ANGLE (Intel, Mesa Intel(R) Xe Graphics (TGL GT2), OpenGL ES 3.2 Mesa 23.2.1-1ubuntu3.1~22.04.4)`


Not sure if this is the Dawn Info line you need.

<img width="1550" height="91" alt="Image" src="https://github.com/user-attachments/assets/f43b294d-b28d-4737-9c95-8191e1540675" />

**Comment by deadprogram** [comment]
That is exactly the line we needed. Thank you.

Your Chrome has `Vulkan: Disabled`, so WebGPU goes through the OpenGL ES
backend of ANGLE in the compatibility mode of Dawn.

```
<Unknown GPU> OpenGLES backend - ANGLE (Intel, Mesa Intel(R) Xe Graphics
(TGL GT2), OpenGL ES 3.2 Mesa 23.2.1-1ubuntu3.1~22.04.4) (Compatibility Mode)
```

That is the cause. On this path the adapter still reports `shader-f16`, so
llama.cpp accepts the device and then computes wrong values on it. Our notes
said such an adapter reports no f16 and that the page falls back to the CPU on
its own. Your machine shows that this is not true, so the note was wrong and we
corrected it.

The card itself is not at fault. The same class of Intel Xe gives correct
values through the Vulkan backend here, so the OpenGL ES path of Dawn is the
part that fails.

Two things:

I will work on a fix. No feature of the adapter tells a device that
operates apart from one that does not, so the loader now measures the device.
It runs one small matrix multiply on the GPU and the same one on the CPU and
compares the results before a model downloads. A device that fails goes away
and the page takes the CPU. A device that operates gives an error near 3e-8
and noise gives an error near 1, so the limit is far from both. This costs a
few milliseconds and needs no model.

Second, a quick test for you, if you want to try it. Close every window of
Chrome and start it this way.

```
google-chrome --enable-features=Vulkan
```

`chrome://gpu` must then say `Vulkan: Enabled`, and the first adapter of Dawn
Info must be a `Vulkan backend` line with the name of your card. We expect
`https://yzma.ai/try/` to give a correct answer on the GPU in that condition.
Please tell us the result, because it confirms that the OpenGL ES backend is
the only part that fails.

**Comment by gc-2412** [comment]
Didn't seem to work. Vulkan enabled.

<img width="579" height="316" alt="Image" src="https://github.com/user-attachments/assets/78e14223-5461-4a60-bdc9-a34cacbbe2f6" />

The dawn info didn't change

<img width="1508" height="85" alt="Image" src="https://github.com/user-attachments/assets/b4b893cb-b911-4684-909f-cfc43ee347ab" />

Garbage output on https://yzma.ai/try/

<img width="878" height="890" alt="Image" src="https://github.com/user-attachments/assets/996bc022-522f-4672-940b-e3450e122ee9" />

**Comment by deadprogram** [comment]
Thank you for the test. The result is useful, even though it did not repair
anything.

The feature status says `Vulkan: Enabled`, but Dawn Info still gives the same
adapter.

```
<Unknown GPU> OpenGLES backend - ANGLE (Intel, Mesa Intel(R) Xe Graphics
(TGL GT2), OpenGL ES 3.2 Mesa 23.2.1-1ubuntu3.1~22.04.4) (Compatibility Mode)
```

So the switch turns Vulkan on for Chrome, and Dawn still has no Vulkan adapter
to give to WebGPU. The page therefore gets the OpenGL ES adapter again, and the
answer is garbage again. Our note said that this switch gives a `Vulkan
backend` line. Your machine shows that it does not always, so we corrected that
note as well.

The conclusion is that your machine has no WebGPU path that operates today. The
one adapter that WebGPU offers is the one that computes wrong values.

This is what the repair does, and it is now in review.

- The page measures the device before it downloads a model.
- Your machine fails that test, so `https://yzma.ai/try/` takes the CPU on its
  own and gives a correct answer.
- You will not need `?mode=cpu` any more.

You lose nothing by this, because the GPU on your machine was never giving a
usable answer.

One optional question, only to explain why Dawn has no Vulkan adapter. It is
not needed for the repair. If `vulkaninfo` is present, please give us the first
lines of this.

```
vulkaninfo --summary
```

If the command is absent, then the Vulkan driver of Mesa is probably not
installed, which would explain the result by itself.

**Comment by gc-2412** [comment]
no such command is installed on my machine

**Comment by deadprogram** [comment]
Probably you need to run:

```
sudo apt install -y mesa-vulkan-drivers vulkan-tools
```
